### ISSUE TITLE
Setting far plane to inf does not work with STDFLOAT_DOUBLE=1 builds

### ISSUE THREAD

**Issue by el-dee** [panda3d/panda3d#987]
## Description

When using Panda3D built with STDFLOAT_DOUBLE=1 setting the far plane of the camera to infinity makes all the objects in the scene to be culled out.

## Steps to Reproduce

* Install Panda3D build with STDFLOAT_DOUBLE=1
* Add to any samples the following line before base.run()

      base.camLens.setNearFar(0.1, math.inf)

* Run the sample, nothing except 2D stuff is rendered.

If one prints the projection matrix, the result is a null matrix

<pre>
0 0 0 0
0 0 0 0
0 0 0 0
0 0 0 0
</pre>

## Environment
 * Operating system: Linux
 * System architecture: x86_64
 * Panda3D version: 1.10.6
 * Installation method: built from source
 * Python version (if using Python): 3.7 and 3.8
 * Compiler (if using C++): N/A

**Comment by ArsThaumaturgis** [comment]
While you don't mention any flickering, could your issue be related to [this one](https://discourse.panda3d.org/t/infinite-far-plane-flickering/25612)? (See in particular the first post, and the one marked as the solution.)

In that case, it seems that under an infinite far-plane, rendered points were turning out to have a depth-value equal to the clear-colour of the depth-buffer, and so under the default depth-test ("less") were being culled.

The solution there was to change the depth-test to be "less or equal"--perhaps it's worth trying that in your case?

**Comment by el-dee** [comment]
> While you don't mention any flickering, could your issue be related to [this one](https://discourse.panda3d.org/t/infinite-far-plane-flickering/25612)? (See in particular the first post, and the one marked as the solution.)
> 
> In that case, it seems that under an infinite far-plane rendered points were turning out to have a depth-value equal to the clear-colour of the depth-buffer, and so under the default depth-test ("less") were being culled.
> 
> The solution there was to change the depth-test to be "less or equal"--perhaps it's worth trying that in your case?

I know about this problem and solution, especially since I'm the author of the post :p

Sadly, this is not the same issue, here the projection matrix is completely wrong (at least that's what the matrix print seems to indicate), and it's not just some part of the scene that end beyond the far plane, but the whole scene which is culled out.

**Comment by rdb** [comment]
The problem does not occur for me, unfortunately.  I modified the Roaming Ralph sample.  Should I use a different sample program?  Did you compile with special options?

**Comment by ArsThaumaturgis** [comment]
> I know about this problem and solution, especially since I'm the author of the post :p

Hah, sorry, I missed that! Well, nevermind then. ^^;

**Comment by el-dee** [comment]
> The problem does not occur for me, unfortunately. I modified the Roaming Ralph sample. Should I use a different sample program?

I mainly tested with the procedural cube, I'm not sure if I tried with Ralph

> Did you compile with special options?

I used the following options :

    --everything --no-gles --no-gles2 --no-egl --use-sse2 --threads 4 --override STDFLOAT_DOUBLE=1 --installer

On a vanilla v1.10.6. The same build without the override works fine.

Looking at the options, I only recently started to use --use-ss2, and I don't think I made a double precision build with it until now, I will try without it and see if there are any differences.

**Comment by el-dee** [comment]
Of course, when quoting the options from what I remembered, I forgot the most important one, and the root cause of the problem --optimize=4

If in the perspectiveLens code I print the output of cinf(fFar), it gives 1 when compiled in simple precision mode, 1 in double precision without optimize and 0 with optimize.

The problem is that the constant used to detect infinite (and also NaN) for double precision value when using code compiled with  finite-math-only (implied by optimize=4) is wrong. I tested with the correct value and it now works, I will submit a PR.
